aptos-move/block-executor/src/scheduler_status.rs

```diff
@@ -0,0 +1,1387 @@
+            // higher incarnation may not have started until the abort finished (here).
+            return Err(code_invariant_error(format!(
+                "Finish abort of incarnation {}, self.next_incarnation_to_try_abort = {}",
+                aborted_incarnation,
+                self.next_incarnation_to_try_abort.load(Ordering::Relaxed),
+            )));
+        }
+
+        {
+            let inner_status = &mut *self.inner_status.lock();
+            if inner_status.already_aborted(aborted_incarnation)
+                || inner_status.never_started_execution(aborted_incarnation)
+            {
+                return Err(code_invariant_error(format!(
+                    "Finish abort of incarnation {}, but inner status {:?}",
+                    aborted_incarnation, inner_status
+                )));
+            }
+
+            match inner_status.status {
+                StatusEnum::Executing => {
+                    inner_status.status = StatusEnum::Aborted;
+                    self.swap_dependency_flag_any(
+                        &[dependency_flags::EXECUTING],
+                        dependency_flags::DEFER,
+                        "finish_abort",
+                    )?;
+                },
+                StatusEnum::Executed => {
+                    self.incarnate(inner_status, new_incarnation, caller_reexecuting);
+                },
+                StatusEnum::PendingScheduling | StatusEnum::Aborted => {
+                    return Err(code_invariant_error(format!(
+                        "Status update to Aborted failed, previous inner status {:?}",
+                        inner_status
+                    )));
+                },
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Adds a stall to the transaction, indicating it has dependencies that might cause re-execution.
+    ///
+    /// When a transaction is stalled, it is removed from the execution queue if in PendingScheduling
+    /// state, or its dependency shortcut is updated from SAFE to DEFER if in Executed state.
+    ///
+    /// # Returns
+    /// - `Ok(true)` if this call changed the state from unstalled to stalled (num_stalls 0→1)
+    /// - `Ok(false)` if the transaction was already stalled or a race condition occurred
+    /// - `Err(PanicError)` if there was an invalid state (e.g., inconsistent status and flag)
+    ///
+    /// # Note
+    /// Each successful add_stall must be balanced by a corresponding remove_stall call.
+    pub(crate) fn add_stall(&self) -> Result<bool, PanicError> {
+        if self.num_stalls.fetch_add(1, Ordering::SeqCst) == 0 {
+            // Acquire write lock for (non-monitor) shortcut modifications.
+            let inner_status = self.inner_status.lock();
+
+            // num_stalls updates are not under the lock, so need to re-check (otherwise
+            // a different remove_stall might have already decremented the count).
+            if self.num_stalls.load(Ordering::Relaxed) == 0 {
```

Comment:
I think it is impossible.

Reply:
oh, "different" was confusing. I believe it depends on the assumption to the caller. If stall and unstall can be concurrent on the caller side (different threads), then this may happen. It might have been the case at some point, I think it won't be the case now, so I will add a comment (that remove_stall must happen after add_stall is fully finished / returns) and remove this.